### DISCUSSION TITLE
setting JVM options

### DISCUSSION THREAD

**Discussion by mghirmay** [wiverson/maven-jpackage-template#63] [General]
hi there! thank your for providing this nice tool. 

i am currently developing a dot.net assembly that needs a certain dotnet-runtime.   I made a JNA wrapper and an JavaFX App which works perfect when i set DOTNET_ROOT on windows via the system-varaible-setting-dialog.

i was wondering if it is posible to use javaOptions to achive this? 
my application cfg-file looks like this:

_[Application]
app.name=javaFxApp
app.version=22.11.031338
app.runtime=$ROOTDIR\runtime
app.identifier=com.so.JDeskApp
app.classpath=$ROOTDIR\app\javaFxApp-1.0.1.jar;.....
app.mainclass=com.so.JDeskApp.MainApplication

[JavaOptions]
-DDOTNET_ROOT=$ROOTDIR/app/dotnetRuntime

[ArgOptions]_

This seems not to work for me. What i am missing?
Any hint would  be very usefull. Thank you.

**Comment by wiverson** [comment]
TBH there are several pieces there, several of which I've never dealt with.

Some observations - setting a DOTNET_ROOT environment variable is not a standard jpackage installer feature AFAIK. So, you could try doing this via customization of the Wix installer, or perhaps by setting the environment variable in your main method before telling the JavaFX app to launch. No idea if that would actually work, tho.

I suspect that the answer to your problem may be to tweak how your JNA wrapper loads the native library (e.g. https://www.baeldung.com/java-jna-dynamic-libraries) to pull it in from the application path. Instead of relying on the DOTNET_ROOT environment variable, I would suggest looking a bit at the way the libraries are getting copied and installed by jpackage and then use that to help sort out step by step which paths to use with the JNA Load.

**Reply by mghirmay** [reply]
thx for the response.

i continued using the approach to set  the enviroment variables on my main class. 

 i added this to my application config file:

[JavaOptions]
-DDOTNET_ROOT=$ROOTDIR\app\dotnetRuntime
-Djna.library.path=$ROOTDIR\app\nativeDeviceApi

Here is my function which i call  set the DOTNET_ROOT.

private static void setSystemProperties() {

        Map<String, String> envReadOnly = System.getenv();
        Map<String, String> envList = new HashMap<>();
        envList.putAll(envReadOnly);
        String dotnetRoot = System.getProperty(DOTNET_ROOT);
        if (dotnetRoot != null) {
            if (isWindows()) {
                System.out.println("windows: set the environment variable: " + dotnetRoot);
                if (!Kernel32.INSTANCE.SetEnvironmentVariable(DOTNET_ROOT, dotnetRoot)) {
                    System.out.println("Unable to set the environemnt variable: " + dotnetRoot);
                } else {
                    envList.put(DOTNET_ROOT, dotnetRoot);
                }
            }
            if (isMac() || isLinux()) {
                System.out.println("max_os+linux: set the environment variable: " + dotnetRoot);
                if (LibC.INSTANCE.setenv(DOTNET_ROOT, dotnetRoot, 1) != 0) {
                    System.out.println("Unable to set the environment variable: " + dotnetRoot);
                } else {
                    envList.put(DOTNET_ROOT, dotnetRoot);
                }
            }
        }

        final String[] output = {""};
        envList.forEach((k, v) -> output[0] += "\n" + k + ":" + v);
        System.out.println(output[0]);
    }

**Reply by wiverson** [reply]
Cool, looks like the next thing you need to do is figure how how to find out where the application is currently installed so you can set the DOTNET_ROOT to the right value.

Some links
https://stackoverflow.com/questions/4032957/how-to-get-the-real-path-of-java-application-at-runtime
https://alvinalexander.com/blog/post/java/java-faq-access-read-system-properties-environment/

Note that the module system puts things in a streamed classloader that may or may not give you what you want. Might be able to find what you are looking for with a Resource load

https://www.geeksforgeeks.org/loading-resources-from-classpath-in-java-with-example/
